docs/05-engineering/git-gitlab-github-bitbucket.md

```diff
@@ -1,29 +1,144 @@
 
 - <https://git-scm.com/>
 - <https://help.github.com/articles/git-and-github-learning-resources/>
 - Getting Git Right <https://www.atlassian.com/git/>
 - Version Control with Git book <http://www.amazon.com/Version-Control-Git-collaborative-development/dp/1449316387>
 - Learn "how git works" at a lower level, and some useful stuff like rebasing (workflow and interactive), reflog, etc. <https://www.youtube.com/watch?v=MYP56QJpDr4>
-- if you want a non-video companion to the first half, <http://git-scm.com/book/en/v2/Git-Internals-Git-Objects> is a great place to start.
+- If you want a non-video companion to the first half, <http://git-scm.com/book/en/v2/Git-Internals-Git-Objects> is a great place to start.
 
-## What's the difference between GitLab, GitHub, and Bitbucket?
+## Git best practices
+
+_Note: your project may include additional or alternate code review practices - please review your project developer/contributor documentation!_
+
+- Atomic commits:
+  - Generally commits should be as small as possible.
+  - The code should be in a functional state after each commit (even if incomplete) - avoid committing code which will won't run at all.
+- Write good commit messages. A good commit message should:
+  - Reference the associated ticket ID using a prefix, e.g. `ABC-123:`.
+  - Have a one line (max 80 char) title/summary of the change, followed by a blank line and then further detail which should:
+    - Explain the reason(s) for the change.
+    - Detail what changed and any consequences of this (on code, data, users, security etc) - include any key terms that would help a future person searching commits.
+    - Often this will include the same detail that you might include in code comments.
+- Feature branches, pull requests (PR) & merge requests (MR):
+  - These should contain a closely related series of commits that solves a single problem - avoid including changes that are unrelated.
+  - Most often a single (well scoped) ticket will be resolved in a single PR/MR.
+    - However it is good to break larger changes over multiple PRs/MRs that each addresses a logical piece of the whole - this can make the code easier to review.
+  - When merging PRs and MRs we prefer a fast-forward/rebase (rather than a merge commit or squash) since this retains a clean and linear commit history.
+- Rebasing commits on a feature branch - we do this for 2 reasons:
+  - To update the feature branch on top of more recent changes in master so that it can be easily merged when ready.
+  - To adjust the commits on the branch with an interactive rebase - for example:
+    - Squashing 2 together that make more sense as one change.
+    - Splitting a large commit into 2+ smaller ones.
+    - Adjusting the wording of commit messages (see above).
+    - Dropping any temporary/scratch commits.
+  - When rebasing a feature branch, it's encouraged to leave commits showing earlier attempts, even if you later found a better approach.
```

Comment:
Squash merges also simplify all of this, you do not have to be as careful with your feature branch, and most of the organization in the feature branch is in service of the reviewer and not for the purpose of your main's branch history staying clean.

Reply:
My rationale for preferring a detailed and reasonably clean history in the main branch is really not so much to optimize for the reviewer, but rather for a future developer N+ years down the line (after PR/MR history is lost due to repo moves etc) who is trying to understand some subtle detail in the code.
I think this is a pretty bike-sheddable thing, and there are reasonable pros and cons for all 3 merge strategies, so totally happy for projects to go the direction that works for them - the intent here is more just to document a default.